vignettes/mocking.Rmd

```diff
@@ -0,0 +1,226 @@
+
+Next we want to check the case where we specify a minimum version, and again we need to make up some version:
+
+```{r}
+test_that("check_installed() checks minimum version", {
+  expect_no_error(check_installed("testthat", "1.0.0"))
+  expect_snapshot(check_installed("testthat", "99.99.999"), error = TRUE)
+})
+```
+
+Again, this is probably save (since I'm unlikely to release 90+ new versions of testthat), but if you look at the snapshot message carefully, you'll notice that it includes the current version of testthat. That means every time a new version of testthat comes out, we'll have to update the snapshot. We could use the `transform` argument to fix that:
+
+```{r}
+test_that("check_installed() checks minimum version", {
+  expect_no_error(check_installed("testthat", "1.0.0"))
+  expect_snapshot(
+    check_installed("testthat", "99.99.999"), 
+    error = TRUE, 
+    transform = function(lines) gsub(packageVersion("testthat"), "<version>", lines)
+  )
+})
+```
+
+But it's starting to feel like we've accumulated a bunch of potentially fragile hacks. So lets see how we could could make these tests more robust with mocking. First we need to add `requireNamspace` and `packageVersion` bindings in our package. This is needed because `requireNamespace` and `packageVersion` are base functions, 
+```{r}
+requireNamespace <- NULL
+packageVersion <- NULL
+```
+
+For the first test, we mock `requireNamespace()` twice, first returning `TRUE`, pretending every package is installed, and then returning `FALSE` pretending that no packages are installed.
+
+```{r}
+test_that("check_installed() checks package is installed", {
+  local_mocked_bindings(requireNamespace = function(...) TRUE)
+  expect_no_error(check_installed("package-name"))
+
+  local_mocked_bindings(requireNamespace = function(...) FALSE)
+  expect_snapshot(check_installed("package-name"), error = TRUE)
+})
+```
+
+For the second test, we mock `requireNamepace()` to return `TRUE`, and then `packageVersion()` to return a fixed version number. Together this simulated that version 2.0.0 of any package is installed and makes our snapshot rely only on the state set within the test.
+
+```{r}
+test_that("check_installed() checks minimum version", {
+  local_mocked_bindings(
+    requireNamespace = function(...) TRUE,
+    packageVersion = function(...) numeric_version("2.0.0")
+  )
+  
+  expect_no_error(check_installed("package-name", "1.0.0"))
+  expect_snapshot(check_installed("package-name", "3.4.5"), error = TRUE)
+})
+```
+
+## Case studies
+
+To give you a bit more experience with mocking, this section looks at a few places where we use mocking throughout the tidyverse. Note that these are all a little complex; this is the nature of mocking: if you can use a simpler technique, you should. So mocking on comes up for otherwise intractable problems.
```

Comment:
```suggestion
To give you a bit more experience with mocking, this section looks at a few places where we use mocking throughout the tidyverse. Note that these are all a little complex; this is the nature of mocking: if you can use a simpler technique, you should. So mocking only comes up for otherwise intractable problems.
```